### DISCUSSION TITLE
Using IOptions for dynamic token insert

### DISCUSSION THREAD

**Discussion by nykez** [Hawxy/Discord.Addons.Hosting#25] [Q&A] [answered]
In the readme, it says that IOptions can be used.

I have tried setting this up without much luck due to the token validator behind the scenes. It is not possible to insert a invalid token or null config into ConfigureDiscordHost. The app is also unable to run without ConfigureDiscordHost (of course). So I don't see how it is possible to enter a bot token via IOptions during runtime..

I can confirm that if I put a valid token in (letting app run normally) and then changing that token via IOptions then it works as intended.

Can a example be shown to get around these issues?

**Comment by Hawxy** [comment, marked as answer]
@nykez Could you explain your scenario a little more? Are you trying to use the `IOptions` configuration builder to fetch the token via dependency injection at startup?

**Reply by nykez** [reply]
Fetch or set the token during startup. In my case, fetching from a dbcontext. If no token is set during startup then the app crashes due to [coming from discord.NET token validator. ](https://github.com/Hawxy/Discord.Addons.Hosting/blob/a275544c0de50b382bdb6d489aa1106fe880a5dd/Discord.Addons.Hosting/DiscordHostBuilderExtensions.cs#L98).

If a token is set during startup then it can be changed to a invalid token during runtime (not that this matters) but I think that is the behavior that should be intended? The app should not crash due to invalid tokens. 

Hopefully this explained it better, thanks.

**Reply by Hawxy** [reply]
You should be able to do something along the lines of: 
```cs
        services.AddOptions<DiscordHostConfiguration>().PostConfigure<MyDbContext>((configuration, dbContext) =>
        {
            configuration.Token = dbContext.Tokens.Single().Token;
        });
```
Validation should not be called until the `IOptions` instance is created during host startup.
